### PR TITLE
Improved ItemStackUtils

### PR DESCRIPTION
**Pull request by Nubebuster**
Fetch uuid from Mojang if server cache doesn't contain valid UUID for given username.

If a player has not played on the server then the offlineplayer will give a random uuid. That results in a steve skin. This patch checks if OfflinePlayer has played before and if not fetch uuid.

### CODE REVIEW

File: plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java [hunk 1]
```diff
@@ -15,25 +15,43 @@
...
+            if (uuidCache.containsKey(username.toLowerCase())) return uuidCache.get(username.toLowerCase());
+
+            String signature = getURLContent(
+                    "https://api.mojang.com/users/profiles/minecraft/" + username);
+            JsonObject profileJsonObject = gson.fromJson(signature, JsonObject.class);
+            String value = profileJsonObject.get("id").getAsString();
+            if (value == null) return null;
+            return uuidCache.put(username.toLowerCase(),
+                    UUID.fromString(value.replaceFirst(
+                            "(\\p{XDigit}{8})(\\p{XDigit}{4})(\\p{XDigit}{4})(\\p{XDigit}{4})(\\p{XDigit}+)",
```
Comment (dlsf) [inline plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java:32]: I don't even know how that Regex is supposed to work but I don't think it does. Also pretty over-complicated
```suggestion
                            "(.{8})(.{4})(.{4})(.{4})(.{12})",
```

If you want - to be ignored use
```suggestion
                            "([^-]{8})([^-]{4})([^-]{4})([^-]{4})([^-]{12})",
```

**Review by dlsf — COMMENTED** [review]
Did you test these changes?

File: plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java (hunk 1 above)
Comment (Nubebuster) [reply plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java:32]: Your regex supports invalid uuids. Not that it matters in this context.

**Comment by Nubebuster** [comment]
> Did you test these changes?

yes

File: plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java (hunk 1 above)
Comment (dlsf) [reply plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java:32]: Sure, but it doesn't matter.

Replace the dots with `[a-f0-9]` so it's also validating if you want that

Comment (dlsf) [reply plugin/src/main/java/com/iridium/iridiumcore/utils/SkinUtils.java:32]: Your regex does the same but is not well readable imo

File: plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java [hunk 2]
```diff
@@ -81,8 +79,18 @@ public static ItemStack makeItem(Item item) {
             if (item.material == XMaterial.PLAYER_HEAD && item.headData != null) {
                 return setHeadData(item.headData, itemstack);
             } else if (item.material == XMaterial.PLAYER_HEAD && item.headOwner != null) {
+                UUID uuid;
```
Comment (PeachesMLG) [inline plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java:82]: I dont really like the look of this code, would something like

```java
OfflinePlayer offlinePlayer = Bukkit.getServer().getOfflinePlayer(item.headOwner);
if(!offlinePlayer.hasPlayedBefore()){
	UUID uuid = SkinUtils.getUUID(item.headOwner);
	if(uuid == null){
		return setHeadData("eyJ0ZXh0dXJlcyI6eyJTS0lOIjp7InVybCI6Imh0dHA6Ly90ZXh0dXJlcy5taW5lY3JhZnQubmV0L3RleHR1cmUvZWI3YWY5ZTQ0MTEyMTdjN2RlOWM2MGFjYmQzYzNmZDY1MTk3ODMzMzJhMWIzYmM1NmZiZmNlOTA3MjFlZjM1In19fQ==", itemstack);
	}
	return setHeadData(SkinUtils.getHeadData(uuid), itemstack);
}else{
        return setHeadData(SkinUtils.getHeadData(offlinePlayer.getUniqueId()), itemstack);
}
```
look better?

File: plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java [hunk 3]
```diff
@@ -81,8 +79,18 @@ public static ItemStack makeItem(Item item) {
...
+                //OfflinePlayer will return a random uuid if the player doesn't exist in server cache
                 OfflinePlayer offlinePlayer = Bukkit.getServer().getOfflinePlayer(item.headOwner);
-                return setHeadData(SkinUtils.getHeadData(offlinePlayer.getUniqueId()), itemstack);
+                if (offlinePlayer.hasPlayedBefore()) {
+                    uuid = offlinePlayer.getUniqueId();
+                } else {
+                    if ((uuid = SkinUtils.getUUID(item.headOwner)) == null) {
```
Comment (PeachesMLG) [inline plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java:88]: Especially this line here, im too dumb to understand it

**Review by PeachesMLG — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java (hunk 3 above)
Comment (Nubebuster) [reply plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java:88]: This line sets the uuid to SkinUtils.getUUID(item.headOwner) and checks if it's null simultaneously

File: plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java (hunk 2 above)
Comment (Nubebuster) [reply plugin/src/main/java/com/iridium/iridiumcore/utils/ItemStackUtils.java]: Whatever you want, it's your plugin :)

**Pull request merged**